### PR TITLE
Normalize alloc-id in tests.

### PR DESCRIPTION
**Pull request by cjgillot**
AllocIds are globally numbered in a rustc invocation. This makes them very sensitive to changes unrelated to what is being tested. This commit normalizes them by renumbering, in order of appearance in the output.

The renumbering allows to keep the identity, that a simple `allocN` wouldn't. This is useful when we have memory dumps.

cc @saethlin 
r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
r=me after rebase and rebless

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -4258,6 +4258,37 @@ impl<'test> TestCx<'test> {
                 V0_BACK_REF_RE.replace_all(&normalized, V0_BACK_REF_PLACEHOLDER).into_owned();
         }
 
+        // AllocId are numbered globally in a compilation session. This can lead to changes
+        // depending on the exact compilation flags and host architecture. Meanwhile, we want
+        // to keep them numbered, to see if the same id appears multiple times.
```
Comment (cjgillot) [inline src/tools/compiletest/src/runtest.rs:4263]: @RalfJung what about this?
https://github.com/rust-lang/rust/pull/116810#discussion_r1361105504

**Comment by RalfJung** [comment]
Ah I didn't realize one PR is based on the other. Please mention that in the PR description. :)

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -4258,6 +4258,37 @@ impl<'test> TestCx<'test> {
...
+            let mut seen_allocs = indexmap::IndexSet::new();
+            normalized = re
+                .replace_all(&normalized, |caps: &Captures<'_>| {
+                    // Use uppercase to distinguish with the non-normalized version.
+                    let mut ret = caps.get(1).unwrap().as_str().to_uppercase();
+                    // Renumber the captured index.
+                    let index = caps.get(2).unwrap().as_str().to_string();
+                    let (index, _) = seen_allocs.insert_full(index);
+                    write!(&mut ret, "{index}").unwrap();
+                    // If we have a tail finishing with `─`, this means pretty-printing.
```
Comment (RalfJung) [inline src/tools/compiletest/src/runtest.rs:4277]: There could be 0 `─` but then a final `╼`.

**Pull request merged**